bridge/engine/RecordingVideoForwarderSendJob.cpp

```diff
@@ -0,0 +1,147 @@
+    EngineMessageListener& mixerManager,
+    size_t endpointIdHash,
+    EngineMixer& mixer)
+    : jobmanager::CountedJob(transport.getJobCounter()),
+      _outboundContext(outboundContext),
+      _senderInboundContext(senderInboundContext),
+      _packet(std::move(packet)),
+      _transport(transport),
+      _extendedSequenceNumber(extendedSequenceNumber),
+      _mixerManager(mixerManager),
+      _endpointIdHash(endpointIdHash),
+      _mixer(mixer)
+{
+    assert(_packet);
+    assert(_packet->getLength() > 0);
+}
+
+void RecordingVideoForwarderSendJob::run()
+{
+    auto rtpHeader = rtp::RtpHeader::fromPacket(*_packet);
+    if (!rtpHeader)
+    {
+        assert(false); // should have been checked multiple times
+        return;
+    }
+
+    if (_outboundContext.rtpMap.format == RtpMap::Format::VP8RTX)
+    {
+        logger::warn("%s rtx packet should not reach rewrite and send. ssrc %u, seq %u",
+            "RecordingVideoForwarderSendJob",
+            _transport.getLoggableId().c_str(),
+            rtpHeader->ssrc.get(),
+            _extendedSequenceNumber);
+        return;
+    }
+
+    if (!_outboundContext.packetCache.isSet())
+    {
+        logger::debug("New ssrc %u seen on %s, sending request to add videoPacketCache to transport",
+            "RecordingVideoForwarderSendJob",
+            _outboundContext.ssrc,
+            _transport.getLoggableId().c_str());
+
+        _outboundContext.packetCache.set(nullptr);
+
+        EngineMessage::Message message(EngineMessage::Type::AllocateRecordingRtpPacketCache);
+        message.command.allocateVideoPacketCache.mixer = &_mixer;
+        message.command.allocateVideoPacketCache.ssrc = _outboundContext.ssrc;
+        message.command.allocateVideoPacketCache.endpointIdHash = _endpointIdHash;
+        _mixerManager.onMessage(std::move(message));
+    }
+
+    const bool isKeyFrame = codec::Vp8Header::isKeyFrame(rtpHeader->getPayload(),
+        codec::Vp8Header::getPayloadDescriptorSize(rtpHeader->getPayload(),
+            _packet->getLength() - rtpHeader->headerLength()));
+
+    const auto ssrc = rtpHeader->ssrc.get();
+    if (ssrc != _outboundContext.originalSsrc)
+    {
+        if (!isKeyFrame)
+        {
+            _outboundContext.needsKeyframe = true;
+            _senderInboundContext.pliScheduler.triggerPli();
+        }
+    }
+
+    if (_outboundContext.needsKeyframe)
+    {
+        if (!isKeyFrame)
+        {
+            // dropping P-frames until key frame appears
+            return;
+        }
+        else
+        {
+            _outboundContext.needsKeyframe = false;
+
+            logger::debug("%s requested key frame from %u on ssrc %u",
+                "RecordingVideoForwarderSendJob",
+                _transport.getLoggableId().c_str(),
+                _senderInboundContext.ssrc,
+                _outboundContext.ssrc);
+        }
+    }
+
+    if (!_outboundContext.shouldSend(rtpHeader->ssrc, _extendedSequenceNumber))
```

Comment:
This is where we check if the packet happens to be too old (512 pkts). It also always allows a new source ssrc to pass through as it means new source stream for rewrite. I guess for recording that rewrite is not actually happening.